### PR TITLE
xds: implement the new v3 and old fallback server xDS API

### PR DESCRIPTION
**Pull request by sanjaypujare**

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/ServerXdsClient.java [hunk 1]
```diff
@@ -99,12 +110,10 @@ private void updateNodeMetadataForListenerRequest(int port) {
     if (node.getMetadata() != null) {
       newMetadata.putAll(node.getMetadata());
     }
-    newMetadata.put("TRAFFICDIRECTOR_PROXYLESS", "1");
-    // TODO(sanjaypujare): eliminate usage of listening_addresses.
-    EnvoyProtoData.Address listeningAddress =
-        new EnvoyProtoData.Address("0.0.0.0", port);
-    node =
-        node.toBuilder().setMetadata(newMetadata).addListeningAddresses(listeningAddress).build();
+    newMetadata.put("TRAFFICDIRECTOR_INBOUND_INTERCEPTION_PORT", "15001");
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/ServerXdsClient.java:113]: 15001 is fixed? Do you have a design for that?

Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/ServerXdsClient.java:113]: This is the "fallback" case where we simulate the Envoy-like xDS API to talk to the TD and this is going to go away so there is no point in having a design. Once the `newServerApi` mode is shown to be tested in production and working then the fallback/legacy code will be removed

File: xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java [hunk 2]
```diff
@@ -68,6 +69,10 @@
   private static final TimeServiceResource timeServiceResource =
       new TimeServiceResource("GrpcServerXdsClient");
 
+  @VisibleForTesting
+  static boolean experimentalNewServerApiEnvVar = Boolean.parseBoolean(
```
[the remark this answers is not in the document]
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java:73]: Unit testing is easier where both modes can be unit tested without trying to set/unset the env variable in the test.

File: xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+/**
+ * Tests for {@link XdsClientImpl2 for server side Listeners using newServerApi}.
+ */
+@RunWith(JUnit4.class)
+public class ServerXdsClientNewServerApiTest {
```
[the remark this answers is not in the document]
Comment (sanjaypujare) [reply xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java:97]: As mentioned elsewhere the class is set up with newServerApi=true in this case for all the tests. In the other file `ServerXdsClientTest` it is invoked with newServerApi=false and this whole file will eventually disappear when newServerApi=true is the only supported mode.

File: xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+        public boolean shouldAccept(Runnable command) {
+          return command.toString().contains(XdsClientImpl2.RpcRetryTask.class.getSimpleName());
+        }
+      };
+  private static final TaskFilter LISTENER_RESOURCE_FETCH_TIMEOUT_TASK_FILTER =
+      new TaskFilter() {
+        @Override
+        public boolean shouldAccept(Runnable command) {
+          return command.toString()
+              .contains(XdsClientImpl2.ListenerResourceFetchTimeoutTask.class.getSimpleName());
```
[the remark this answers is not in the document]
Comment (sanjaypujare) [reply xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java:119]: good catch. FIxed.

File: xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+  private static final String TYPE_URL_HCM =
+      "type.googleapis.com/"
+          + "envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager";
+
+  private static final Node NODE = Node.newBuilder().build();
+  private static final TaskFilter RPC_RETRY_TASK_FILTER =
+      new TaskFilter() {
+        @Override
+        public boolean shouldAccept(Runnable command) {
+          return command.toString().contains(XdsClientImpl2.RpcRetryTask.class.getSimpleName());
```
[the remark this answers is not in the document]
Comment (sanjaypujare) [reply xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java:111]: good catch. Fixed

**Comment by sanjaypujare** [comment]
@voidzcy I have addressed all the comments.

File: xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java:97]: This is a very verbose way though, with a lot of duplicates. Gating an unreleased feature with environment variable should be pure implementation detail and removing the environment variable should not require change how the class is used/created. Also, you shouldn't re-test logics not related to the feature the env variable is gating, you'd at most need one or two extra test cases verifying the xDS requests it sends are based on the environment value. Otherwise, if you need to change some of the behaviors unrelated to what the env variable is gating for, you will need to change the corresponding tests in both places.

[the remark this answers is not in the document]
Comment (sanjaypujare) [reply xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java:97]: This is not different from the duplication of tests in `ServerXdsClientTest` and `XdsClientImplTestForListener`. In that case the tests are the same but the class being tested is different (`ServerXdsClient` and `XdsClientImpl2`). In this PR the class is the same but constructed with newServerApi=true vs false.

Also for every test the DiscoveryRequest is different (it has a resource name vs empty resource in the old test). Once the newServerApi becomes the norm the old code will be removed and any code related to `newServerApi==false` will be removed which is easier with the current structure. I prefer this structure to make it easy to just remove the old code.

[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java:97]: `XdsClientImplTestForListener` is intended to be deleted. It has been not because the client side is still using `XdsClientImpl2`. Once all usages of `XdsClientImpl2` are migrated, it will be deleted along with its tests (including `XdsClientImplTestForListener`).

Regarding duplicates (tests for stream retry) in `ClientXdsClientTest` and `ServerXdsClient`, they are run by completely different codepaths and developed by different authors, so it's acceptable to have two sets of tests for the purpose of easy collaboration. Ideally, we should have a way to cover stream retry logics only once.

But here, you are writing two complete sets of tests for the same piece of logic. You will need to maintain both sets together when there are new changes. Since this is your own code and no one else uses it, I am fine if you are ok with the maintenance burden or you feel this is easier for you.

**Review by voidzcy — APPROVED** [review]

File: xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sanjaypujare) [reply xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java:97]: Believe me, this will make it easy for us to fall back to the old API in case the new API doesn't work with the control plane. And if it works it shouldn't take any time to remove the old code.
